quinn-proto/src/connection/streams.rs

```diff
@@ -484,17 +491,18 @@ impl Streams {
         // MAX_DATA
         if pending.max_data && buf.len() + 9 < max_size {
             pending.max_data = false;
-            // We assume that if local_max_data is > VarInt::MAX, it must be due to an
-            // astronomically high configured initial credit, so we don't need to worry about the
-            // sender actually getting blocked somewhere between the initial credit and the current
-            // one.
-            if let Ok(max) = VarInt::from_u64(self.local_max_data) {
-                trace!(value = self.local_max_data, "MAX_DATA");
-                sent.max_data = true;
-                buf.write(frame::Type::MAX_DATA);
-                buf.write(max);
-                stats.max_data += 1;
-            }
+
+            // `local_max_data` can grow bigger than `VarInt`.
+            // For transmission inside Quic frames we need to clamp it to the
+            // maximum allowed `VarInt` size.
+            let max = core::cmp::min(self.local_max_data, VarInt::MAX.into_inner());
```

Comment:
`let max = VarInt::try_from(self.local_max_data).unwrap_or(VarInt::MAX);` and replace `write_var` with `write` for stronger typing

Reply:
Makes sense - I actually had an error first due to not changing the write call to `write_var`.

But if we immediately move to `VarInt`, we have to convert it back to `u64` for the `trace!` and `record_sent_max_data` (although I could change the latter to take `VarInt`) 🤷‍♂️

Guess I'm fine with either approach

Reply:
I'd lean towards making the change, including updating `record_sent_max_data`, since indeed we can never send an unrepresentable value. I've been trying to gradually move towards making more internal use of `VarInt` project-wide to reduce opportunity for error.